packages/records/lib/models/records.ts

```diff
@@ -355,49 +338,61 @@ export async function markNonCurrentGenerationRecordsAsDeleted(connectionId: num
 }
 
 /**
- * Compute Added Keys
- * @desc for any incoming payload use the provided unique to check against the rows
- * in the database and return the keys that are not in the database
- *
+ * getUpdatedKeys
+ * @desc returns a list of the keys that exist in the records tables but have a different data_hash
  */
-async function getAddedKeys(response: FormattedRecord[], connectionId: number, model: string): Promise<string[]> {
-    const keys: string[] = response.map((record: FormattedRecord) => getUniqueId(record));
+async function getUpdatedKeys(records: FormattedRecord[], connectionId: number, model: string): Promise<string[]> {
+    const keys: string[] = records.map((record: FormattedRecord) => getUniqueId(record));
+    const keysWithHash: [string, string][] = records.map((record: FormattedRecord) => [getUniqueId(record), record.data_hash]);
 
-    const knownKeys: string[] = (await db
+    const rowsToUpdate = (await db
         .from(RECORDS_TABLE)
+        .pluck('external_id')
         .where({
             connection_id: connectionId,
-            model,
-            deleted_at: null
+            model
         })
         .whereIn('external_id', keys)
-        .pluck('external_id')) as unknown as string[];
+        .whereNotIn(['external_id', 'data_hash'], keysWithHash)) as unknown as string[];
 
-    return keys?.filter((key: string) => !knownKeys.includes(key));
+    return rowsToUpdate;
 }
 
-/**
- * Get Updated Keys
- * @desc generate an array of the keys that exist in the database and also in
- * the incoming payload that will be used to update the database.
- * Compare using the data_hash key
- *
- */
-async function getUpdatedKeys(response: FormattedRecord[], connectionId: number, model: string): Promise<string[]> {
-    const keys: string[] = response.map((record: FormattedRecord) => getUniqueId(record));
-    const keysWithHash: [string, string][] = response.map((record: FormattedRecord) => [getUniqueId(record), record.data_hash]);
-
-    const rowsToUpdate = (await db
+async function getUpsertSummary(
+    records: FormattedRecord[],
+    connectionId: number,
+    model: string,
+    nonUniqueKeys: string[],
+    softDelete: boolean
+): Promise<UpsertSummary> {
+    const keys: string[] = records.map((record: FormattedRecord) => getUniqueId(record));
+    const nonDeletedKeys: string[] = await db
         .from(RECORDS_TABLE)
-        .pluck('external_id')
         .where({
             connection_id: connectionId,
-            model
+            model,
+            deleted_at: null
         })
         .whereIn('external_id', keys)
-        .whereNotIn(['external_id', 'data_hash'], keysWithHash)) as unknown as string[];
+        .pluck('external_id');
 
-    return rowsToUpdate;
+    if (softDelete) {
+        return {
+            addedKeys: [],
+            updatedKeys: [],
+            deletedKeys: nonDeletedKeys,
```

Comment:
this part is confusing, I don't get how a select `deleted_at:null` can be used for deletedKeys

Reply:
because this query happens before the deletion so all all known records that are not deleted yet (`select ... from ... where deleted is null and external_id in (list of keys)`) will then be deleted